### DISCUSSION TITLE
[Proposal]: calling static generic method in generic class without the generic type of the class like java

### DISCUSSION THREAD

**Discussion by qhc99** [dotnet/csharplang#4375] [Language Ideas] [answered]
# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary
A small grammer improvement. It doesn't have any effect to fill a type in a generic class when calling its' generic static method since the method is generic itself(class generic type doesn't equal method generic type or else it will raise warning CS0693). Calling the method without specifiing a type for the class should be legal

## Motivation
[motivation]: #motivation
more elegant grammer

## Detailed design
[design]: #detailed-design
see the calling of "MethodA", it is very easy to understand this case.
```
using System;

namespace CSharpScripts
{
    public static class Program
    { 
        public static void Main(string[] args)
        {
            // there is no need to sepecify a type for ClassA 
            // in fact any type can be sepecified here(e.g. int, double, ...).
            // but ClassA.MethodA(1) won't compile
            // java dose't have this problem
            ClassA<object>.MethodA(1);
        }

        public class ClassA<T>
        {
            public static void MethodA<TI>(TI ti)
            {
                Console.WriteLine(ti);
            }
        }
        
    }
}

```
also note the case in "using static"
```
using System;
// the design should be "using static CSharpScripts.Program.ClassA";
using static CSharpScripts.Program.ClassA<bool>;

namespace CSharpScripts
{
    public static class Program
    { 
        public static void Main(string[] args)
        {
            MethodA(1);
        }

        public class ClassA<T>
        {
            public static void MethodA<TI>(TI ti)
            {
                Console.WriteLine(ti);
            }
        }
        
    }
}
```
these two pieces of code run normally in my computer.

## Drawbacks
[drawbacks]: #drawbacks
Java has this feature and it is definitely not harmful

## Alternatives
[alternatives]: #alternatives
Don't fix it and specify the type infered by the method. But it dosen't have any effect on program and it maybe confusing when someone specify a type that doesn't equal with the infered type. 

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by YairHalberstadt** [comment, marked as answer]
Java doesn't have generics at runtime and so all static methods on a generic class are actually the same. In C# this is not the case - theoretically a different method is compiled for every type parameter used in static method (ignoring done optimisations to share generated code) and so the type parameter must be set.

E.g, this is legal in c#:

```csharp
static class A<T> where T : struct
{
     static object New() => new T();
}

A.New(); //what would this return?
```

**Reply by qhc99** [reply]
I've understood it, thanks!
